src/plugins/view_control/ViewControl.cc

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <ignition/common/MouseEvent.hh>
+
+#include <ignition/gui/Application.hh>
+#include <ignition/gui/GuiEvents.hh>
+#include <ignition/gui/MainWindow.hh>
+
+#include <ignition/plugin/Register.hh>
+
+#include <ignition/rendering/Camera.hh>
+#include <ignition/rendering/OrbitViewController.hh>
+#include <ignition/rendering/RenderingIface.hh>
+#include <ignition/rendering/RayQuery.hh>
+
+#include "ViewControl.hh"
+
+/// \brief Private data class for ViewControl
+class ignition::gui::plugins::ViewControlPrivate
+{
+  /// \brief
+  public: void OnRender();
+
+  math::Vector3d ScreenToScene(
+      const math::Vector2i &_screenPos) const;
+
+  /// \brief Flag to indicate if mouse event is dirty
+  public: bool mouseDirty = false;
+
+  public: bool blockOrbit = false;
+
+  /// \brief Mouse event
+  public: common::MouseEvent mouseEvent;
+
+  /// \brief Mouse move distance since last event.
+  public: math::Vector2d drag;
+
+  /// \brief User camera
+  public: rendering::CameraPtr camera;
+
+  /// \brief View control focus target
+  public: math::Vector3d target;
+
+  /// \brief Camera orbit controller
+  public: rendering::OrbitViewController viewControl;
+
+  /// \brief Ray query for mouse clicks
+  public: rendering::RayQueryPtr rayQuery;
+
+  //// \brief Pointer to the rendering scene
+  public: rendering::ScenePtr scene = nullptr;
+};
+
+using namespace ignition;
+using namespace gui;
+using namespace plugins;
+
+/////////////////////////////////////////////////
+void ViewControlPrivate::OnRender()
+{
+  if (!this->scene)
+  {
+    this->scene = rendering::sceneFromFirstRenderEngine();
+    if (!this->scene)
+      return;
+
+    this->camera = std::dynamic_pointer_cast<rendering::Camera>(
+      this->scene->SensorByName("Scene3DCamera"));
```

Comment:
I think that hardcoding the name of the camera may be a bit limiting. Can we assume that the first camera is the user camera? Like the screenshot plugin does:

https://github.com/ignitionrobotics/ign-gui/blob/d509ab956e50527daa6005345606b77c4f980bbf/src/plugins/screenshot/Screenshot.cc#L187-L198

Reply:
With the recent changes in the PR of running in the process this may be not true. I included this change, I think all cameras except the one is moving should have a name similar to `scene::camera(XXXX)` otherwise should be named like `model_name::link_name::camera_name`

https://github.com/ignitionrobotics/ign-gui/blob/f3de5463e94e01ea9bebf640d1034be7356e6898/src/plugins/interactive_view_control/InteractiveViewControl.cc#L82-L95

Reply:
> With the recent changes in the PR of running in the process this may be not true

Yeah I see your point. The problem is that the scene of `Scene3D` or `MinimalScene` may not be called `scene` if the user passes a custom name through XML's `<scene>` element. I ran into this while writing a test for #226 which had a different scene name.

So I think we need a different way of identifying the user camera. On Gazebo classic there was a separate class for the user camera, so that was easy. On Ignition Rendering I don't think there's any distinction.

Another thing to keep in mind is that in the future we may want to restore support for multiple user cameras in the same scene, and these plugins that attach to a camera may want to choose which camera to attach to ([like we had in `ign-gui0`](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gui/pull-requests/74/page/1)).

Once https://github.com/ignitionrobotics/ign-rendering/issues/58 is tackled, maybe we could store a different property into the user camera(s)? @iche033 , do you have other ideas on how to identify if a camera is on the GUI or a sensor?

Reply:
PR https://github.com/ignitionrobotics/ign-rendering/pull/358